src/ghci/manager.rs

```diff
@@ -51,29 +70,62 @@ pub async fn run_ghci(
     }
 
     let ghci = Arc::new(Mutex::new(ghci));
+    // The event to respond to. If we interrupt a reload, we may begin the loop with `Some(_)` in
+    // here.
+    let mut maybe_event = None;
     loop {
-        // Wait for filesystem events.
-        let event = tokio::select! {
-            _ = handle.on_shutdown_requested() => {
-                ghci.lock().await.stop().await.wrap_err("Failed to quit ghci")?;
-                break;
-            }
-            ret = receiver.recv() => {
-                ret.ok_or_else(|| miette!("ghci event channel closed"))?
+        let mut event = match maybe_event.take() {
+            Some(event) => event,
+            None => {
+                // If we don't already have an event to respond to, wait for filesystem events.
+                let event = tokio::select! {
+                    _ = handle.on_shutdown_requested() => {
+                        ghci.lock().await.stop().await.wrap_err("Failed to quit ghci")?;
+                        break;
+                    }
+                    ret = receiver.recv() => {
+                        ret.ok_or_else(|| miette!("ghci event channel closed"))?
+                    }
+                };
+                tracing::debug!(?event, "Received ghci event from watcher");
+                event
             }
         };
-        tracing::debug!(?event, "Received ghci event from watcher");
 
+        // This channel notifies us what kind of reload is triggered, which we can use to inform
+        // our decision to interrupt the reload or not.
+        let (reload_sender, reload_receiver) = oneshot::channel();
         // Dispatch the event. We spawn it into a new task so it can run in parallel to any
         // shutdown requests.
-        let mut task = Box::pin(tokio::task::spawn(dispatch(ghci.clone(), event)));
+        let mut task = Box::pin(tokio::task::spawn(dispatch(
+            ghci.clone(),
+            event.clone(),
+            reload_sender,
+        )));
         tokio::select! {
             _ = handle.on_shutdown_requested() => {
                 // Cancel any in-progress reloads. This releases the lock so we don't block here.
                 task.abort();
                 ghci.lock().await.stop().await.wrap_err("Failed to quit ghci")?;
                 break;
             }
+            Some(new_event) = receiver.recv() => {
```

Comment:
Might be worth explicitly matching on the `Reload` constructor of `GhciEvent` so that if you later add more constructors this will become a non-exhaustive pattern match.